File: src/ashura/health/kafka/producer.py
import logging
from django.conf import settings
from confluent_kafka import Producer
from .serializer import serialize_message

logger = logging.getLogger("ashura_kafka_producer")

class Singleton(type):
    _instances = {}
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]


class KafkaProducer(metaclass=Singleton):
    def __init__(self):
        self.producer = Producer({'bootstrap.servers': settings.KAFKA_BOOTSTRAP_SERVERS})

    def send_message(self, topic, message):
        serialized_message = message

        try:
            self.producer.produce(topic, serialized_message.encode('utf-8'))
            self.producer.flush()
            logger.info(f'Message sent to Kafka: {serialized_message}')
        except Exception as e:
            logger.error(f'Error occurred while producing to Kafka: {e}')

File: src/indra/src/kafka/producer.ts
import { logger } from "@cashoco/common";
import { Kafka, Producer, ProducerRecord } from "kafkajs";

const KAFKA_BROKER :string = process.env.KAFKA_BROKER!;

export class KafkaProducer{
    
    private static instance: KafkaProducer;
    private producer: Producer;
    
    constructor()
    {
        const kafka = new Kafka({
            brokers:[KAFKA_BROKER]
        })
        this.producer=kafka.producer()      
    }

    public static getInstance(): KafkaProducer {
        if(!KafkaProducer.instance){
            KafkaProducer.instance= new KafkaProducer();
        }
        return KafkaProducer.instance;
  }
    
    public sendMessage = async ( topic:string, message: string)=>{
        await this.producer.connect()

        const payload: ProducerRecord ={
            topic,
            messages:[{value:message}],
        }
        
        await this.producer.send(payload)
            .then(()=>{
                logger.info(`Message sent to Kafka: ${message}`);
            })
            .catch(err=>{ 
                logger.error(`Error occurred while producing to Kafka: ${err}`);
            })
   }
}